test-framework/junit5/src/main/java/io/quarkus/test/junit/classloading/FacadeClassLoader.java

```diff
@@ -546,10 +546,19 @@ private QuarkusClassLoader getOrCreateBaseClassLoader(String key, Class<?> requi
     private StartupAction getOrCreateRuntimeClassLoader(String key, Class<?> requiredTestClass, Class<?> profile)
             throws ClassNotFoundException, NoSuchMethodException, InvocationTargetException, InstantiationException,
             IllegalAccessException, AppModelResolverException, BootstrapException, IOException {
+        ClassLoader old = Thread.currentThread().getContextClassLoader();
         CuratedApplication curatedApplication = getOrCreateCuratedApplication(key, requiredTestClass);
+
+        // Before interacting with the profiles, set the TCCL to one which is not the FacadeClassloader
+        // This could also go in AppMakerHelper.setExtraProperties, but then it affects more code paths
+        if (profile != null) {
+            Thread.currentThread().setContextClassLoader(profile.getClassLoader());
+        }
         StartupAction startupAction = AppMakerHelper.getStartupAction(requiredTestClass,
                 curatedApplication, profile);
 
+        Thread.currentThread().setContextClassLoader(old);
```

Comment:
Should we do it in a `finally` block?

Reply:
Yes, good idea. I had it in a finally block and took it out because the code was easier to read with less indentation, and I decided if an exception happened everything was terrible and so it wouldn't matter if the TCCL was wrong. Will update.

Reply:
Oh, I remember why I did it without the finally block. If I want to minimise the scope of when we're using the other classloader, and set it back before ` QuarkusClassLoader loader = startupAction.getClassLoader()`,  I'd have to move the `startupAction` out of the scope of the try, and initialise it to null. It adds a bit of cognitive load about why there's no NPE risk.